### PR TITLE
EZP-32258: OAuth2 and log in via external service

### PR DESCRIPTION
**Pull request by DominikaK**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | https://issues.ibexa.co/browse/EZP-32258
| Versions      | 3.3, master

To do: 
- [x] Add new Twig functions to reference

### CODE REVIEW

File: code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+knpu_oauth2_client:
+    clients:
+        # configure your clients as described here: https://github.com/knpuniversity/oauth2-client-bundle#configuration
+        google:
+            type: google
+           client_id: '%env(OAUTH_GOOGLE_CLIENT_ID)%'
```
Comment (kmadejski) [inline code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml:6]: These two `client_*` lines have the wrong indentation that will cause an exception.

File: code_samples/user_management/oauth_google/config/services.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    # controllers are imported separately to make sure services can be injected
+    # as action arguments even if you don't extend any base controller class
+    App\Controller\:
+        resource: '../src/Controller/'
+        tags: ['controller.service_arguments']
+
+    # add more service definitions when explicit configuration is needed
+    # please note that last definitions always *replace* previous ones
+
+    App\OAuth\GoogleResourceOwnerMapper:
```
Comment (kmadejski) [inline code_samples/user_management/oauth_google/config/services.yaml:34]: I think we need only lines from 34 to the end and actually, it most likely should be moved to a separated file, WDYT @adamwojs?

File: docs/guide/twig_functions_reference.md [hunk 3]
```diff
@@ -30,6 +30,8 @@ In addition to the [native functions provided by Twig](http://twig.sensiolabs.or
 |[`ez_render_location_query`](#ez_render_location_query)|Renders results of a non-content related Location query.|
 |[`ez_render_field`](#ez_render_field)|Displays a Content item's Field value, taking advantage of the template block exposed by the Field Type used.|
 |[`ez_urlalias`](#ez_urlalias)|It is a special route name for generating URLs for a Location from the given parameters.|
+|[`ibexa_oauth2_connect_path`](#ibexa_oauth2_connect_path)|Generates a relative path for the given OAuth2 route|
```
Comment (kmadejski) [inline docs/guide/twig_functions_reference.md:33]:
```suggestion
|[`ibexa_oauth2_connect_path`](#ibexa_oauth2_connect_path)|Generates a relative path for the given OAuth2 client|
```

File: docs/guide/twig_functions_reference.md [hunk 4]
```diff
@@ -30,6 +30,8 @@ In addition to the [native functions provided by Twig](http://twig.sensiolabs.or
...
+|[`ibexa_oauth2_connect_url`](#ibexa_oauth2_connect_url)|Generates an absolute URL for the given OAuth2 route|
```
Comment (kmadejski) [inline docs/guide/twig_functions_reference.md:34]:
```suggestion
|[`ibexa_oauth2_connect_url`](#ibexa_oauth2_connect_url)|Generates an absolute URL for the given OAuth2 client|
```

**Review by kmadejski — CHANGES_REQUESTED** [review]

File: code_samples/user_management/oauth_google/config/services.yaml (hunk 2 above)
Comment (DominikaK) [reply code_samples/user_management/oauth_google/config/services.yaml:34]: Only lines 34+ will be embedded in the doc, see here: https://github.com/ezsystems/developer-documentation/pull/1304/files#diff-8fcf856c9fcd93fd5bbc9d15cf7c49a16463d30067a431d946618eb1d4c0bbd0R51
The full file is included for testing purposes, but not rendered.

File: code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+           client_secret: '%env(OAUTH_GOOGLE_CLIENT_SECRET)%'
```
Comment (DominikaK) [inline code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml:7]:
```suggestion
            client_id: '%env(OAUTH_GOOGLE_CLIENT_ID)%'
            client_secret: '%env(OAUTH_GOOGLE_CLIENT_SECRET)%'
```

File: code_samples/user_management/oauth_google/config/services.yaml (hunk 2 above)
Comment (kmadejski) [reply code_samples/user_management/oauth_google/config/services.yaml:34]: Oh, sorry, I didn't notice it. All good 👍

**Review by kmadejski — APPROVED** [review]

File: code_samples/user_management/oauth_google/src/OAuth/GoogleResourceOwnerMapper.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
+<?php
+declare(strict_types=1);
```
Comment (adamwojs) [inline code_samples/user_management/oauth_google/src/OAuth/GoogleResourceOwnerMapper.php:2]: Nitpick: missing new line.

```suggestion

declare(strict_types=1);
```

**Review by adamwojs — APPROVED** [review]

File: docs/guide/user_management/login_via_external_service.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+Configure the OAuth2 client in `config/packages/knpu_oauth2_client.yaml`:
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml') =]]
+```
+
+## Enable OAuth authentication
+
+Enable OAuth2 authentication via Google for the `site` SiteAccess:
```
Comment (dabrt) [inline docs/guide/user_management/login_via_external_service.md:17]:
```suggestion
Enable OAuth2 authentication through Google for the `site` SiteAccess:
```

File: docs/guide/user_management/login_via_external_service.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/packages/oauth.yaml') =]]
+```
+
+## Configure firewall
+
+Add the `Ibexa\Platform\OAuth2Client\Security\Authenticator\OAuth2Authenticator` guard authenticator
+to your firewall configuration in `config/packages/security.yaml`
+and ensure that the `ibexa.oauth2.connect` route is accessible by anonymous user:
```
Comment (dabrt) [inline docs/guide/user_management/login_via_external_service.md:27]:
```suggestion
and ensure that the `ibexa.oauth2.connect` route is accessible by an anonymous user:
```

File: docs/guide/user_management/login_via_external_service.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+The mapper loads a user (line 50) or creates a new one (line 60),
+based on the information from `resourceOwner`, that is the OAuth provider.
+
+The new user name is set with a `google:` prefix (lines 18, 105), to avoid conflicts with users registered in a regular way.
+
+``` php hl_lines="18 50 60 105"
+[[= include_file('code_samples/user_management/oauth_google/src/OAuth/GoogleResourceOwnerMapper.php') =]]
+```
+
+Configure the service using the `ibexa.oauth2_client.resource_owner_mapper` tag:
```
Comment (dabrt) [inline docs/guide/user_management/login_via_external_service.md:48]:
```suggestion
Configure the service by using the `ibexa.oauth2_client.resource_owner_mapper` tag:
```

File: docs/guide/user_management/login_via_external_service.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/services.yaml', 33, 36) =]]
+```
+
+## Add template
+
+To add a "Log in via Google" button to your Back Office login form, create the following template file
```
Comment (dabrt) [inline docs/guide/user_management/login_via_external_service.md:56]:
```suggestion
To add a "Log in with Google" button to your Back Office login form, create the following template file
```

File: docs/guide/user_management/login_via_external_service.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+in `templates/themes/admin/account/login/oauth2_login.html.twig`:
+
+``` html+twig
+[[= include_file('code_samples/user_management/oauth_google/templates/themes/admin/account/login/oauth2_login.html.twig') =]]
+```
+
+Finally, add the template to the login form using the `login-form-after` [component](../../extending/custom_components.md):
```
Comment (dabrt) [inline docs/guide/user_management/login_via_external_service.md:63]:
```suggestion
Finally, add the template to the login form by using the `login-form-after` [component](../../extending/custom_components.md):
```

File: docs/guide/user_management/login_via_external_service.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/services.yaml', 37, 43) =]]
+```
+
+![Log in to the Back Office via Google](../img/log_in_via_google.png)
```
Comment (dabrt) [inline docs/guide/user_management/login_via_external_service.md:69]:
```suggestion
![Log in to the Back Office with Google](../img/log_in_via_google.png)
```

File: docs/guide/user_management/oauth.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+You can use OAuth2 authentication to securely connect to external services.
+
+[[= product_name =]] uses an integration with [`knpuniversity/oauth2-client-bundle`](https://github.com/knpuniversity/oauth2-client-bundle)
+to provide OAuth2 authentication.
+
+To enable OAuth2, you need to:
+
+- [enable and configure an OAuth2 provider](#oauth2-provider-configuration)
+- [add a guard authentication to your firewall configuration](#firewall-configuration)
+- [select one of the existing resource owner mappers, or implement your own](#resource-owner-mappers)
```
Comment (dabrt) [inline docs/guide/user_management/oauth.md:12]:
```suggestion
- [select one of the existing resource owner mappers, or implement your own one](#resource-owner-mappers)
```

File: docs/guide/user_management/oauth.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+## OAuth2 provider configuration
+
+To configure the OAuth2 provider, add it under the `oauth2` key in SiteAccess-aware configuration, for example:
+
+``` yaml
+[[= include_file('code_samples/user_management/oauth_google/config/packages/oauth.yaml') =]]
+```
+
+Details of the configuration depend on the OAuth2 provider you want to use.
```
Comment (dabrt) [inline docs/guide/user_management/oauth.md:22]:
```suggestion
Details of the configuration depend on the OAuth2 provider that you want to use.
```

File: docs/guide/user_management/oauth.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+To see example configurations for different providers,
```
Comment (dabrt) [inline docs/guide/user_management/oauth.md:23]:
```suggestion
For sample configurations for different providers,
```

File: docs/guide/user_management/oauth.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+check [`knpuniversity/oauth2-client-bundle` configuration](https://github.com/knpuniversity/oauth2-client-bundle#configuration).
```
Comment (dabrt) [inline docs/guide/user_management/oauth.md:24]:
```suggestion
see [`knpuniversity/oauth2-client-bundle` configuration](https://github.com/knpuniversity/oauth2-client-bundle#configuration).
```

File: docs/guide/user_management/oauth.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+- `ResourceOwnerToExistingUserMapper` is the base class that is extended by the following mappers:
+    - `ResourceOwnerIdToUserMapper` does not create a new user, but loads a user (resource owner) based on their identifier.
+    - `ResourceOwnerEmailToUserMapper` does not create a new user, but loads a user (resource owner) based on their email.
+- `ResourceOwnerToExistingOrNewUserMapper` checks if the user exists and loads them if they do.
+If they do not, the mapper creates a new user in the Repository.
+
+To use `ResourceOwnerToExistingOrNewUserMapper` you need to extend it in your custom mapper.
+
+See [Adding login via external service](login_via_external_service.md) for an example of creating a mapper
+extending `ResourceOwnerToExistingOrNewUserMapper`.
```
Comment (dabrt) [inline docs/guide/user_management/oauth.md:51]:
```suggestion
that extends `ResourceOwnerToExistingOrNewUserMapper`.
```

File: docs/guide/user_management/oauth.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+- `ResourceOwnerToExistingUserMapper` is the base class that is extended by the following mappers:
+    - `ResourceOwnerIdToUserMapper` does not create a new user, but loads a user (resource owner) based on their identifier.
+    - `ResourceOwnerEmailToUserMapper` does not create a new user, but loads a user (resource owner) based on their email.
+- `ResourceOwnerToExistingOrNewUserMapper` checks if the user exists and loads them if they do.
+If they do not, the mapper creates a new user in the Repository.
+
+To use `ResourceOwnerToExistingOrNewUserMapper` you need to extend it in your custom mapper.
+
+See [Adding login via external service](login_via_external_service.md) for an example of creating a mapper
```
Comment (dabrt) [inline docs/guide/user_management/oauth.md:50]:
```suggestion
See [Adding login through external service](login_via_external_service.md) for an example of creating a mapper
```

File: mkdocs.yml [hunk 19]
```diff
@@ -177,6 +177,8 @@ nav:
         - User management:
             - 'User management': 'guide/user_management/user_management.md'
             - 'Login and registration': 'guide/user_management/login_and_registration.md'
+            - 'OAuth authentication': 'guide/user_management/oauth.md'
+            - 'Adding login via external service': 'guide/user_management/login_via_external_service.md'
```
Comment (dabrt) [inline mkdocs.yml:181]:
```suggestion
            - 'Adding login through external service': 'guide/user_management/login_via_external_service.md'
```

**Review by dabrt — APPROVED** [review]

File: code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml [hunk 20]
```diff
@@ -0,0 +1,11 @@
+knpu_oauth2_client:
+    clients:
+        # configure your clients as described here: https://github.com/knpuniversity/oauth2-client-bundle#configuration
```
[the remark this answers is not in the document]
Comment (DominikaK) [reply code_samples/user_management/oauth_google/config/packages/knpu_oauth2_client.yaml:3]: Nope, that's a good comment, if it's a code file created for the doc, and not just quoted from the product code.

File: docs/guide/twig_functions_reference.md [hunk 21]
```diff
@@ -669,3 +671,35 @@ For a Location alias set up a 301 redirect to the Location's current URL when:
...
+
+#### Prototype and Arguments
+
+`ibexa_oauth2_connect_url ( string $identifier [, array $parameters, bool $schemeRelative ] ): string`
+
+|Argument name|Type|Description|
+|------|------|------|
+|`identifier`|string|Identifier of the OAuth connection|
+|`parameters`|`array`|Route parameters|
+|`schemeRelative`|`boolean`|Whether to generate a relative URL|
```
Comment (DominikaK) [inline docs/guide/twig_functions_reference.md:705]:
```suggestion
|`schemeRelative`|`boolean`|Specifies whether to generate a relative URL|
```
Re: `boolean` spalling: https://www.php.net/manual/en/language.types.boolean.php

File: docs/guide/user_management/login_via_external_service.md (hunk 10 above)
Comment (DominikaK) [inline docs/guide/user_management/login_via_external_service.md:56]:
```suggestion
To add a **Log in with Google** button to your Back Office login form, create the following template file
```

File: docs/guide/twig_functions_reference.md (hunk 21 above)
Comment (dabrt) [reply docs/guide/twig_functions_reference.md]: Nevertheless, I disagree. Boolean is an adjective that derives from a surname and as such, should be capitalized.
We are not talking about code here - in code developers can use whatever syntax requires them to use. We are describing the nature of a variable in the documentation (we might as well get rid of the code delimiters in this column, as we are not quoting code) and I'd rather stick to what the following sources dictate:
https://www.merriam-webster.com/dictionary/Boolean
https://www.ldoceonline.com/dictionary/boolean
https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/boolean

File: docs/guide/user_management/login_via_external_service.md (hunk 10 above)
Comment (dabrt) [reply docs/guide/user_management/login_via_external_service.md]: Missed that

**Pull request merged**